cmd/testperformance.go

```diff
@@ -0,0 +1,614 @@
+
+	if !cfg.Quiet {
+		err = writeResultToWriter(res, w)
+		if err != nil {
+			return err
+		}
+	}
+
+	if cfg.OutputToml != "" {
+		err = writeResultToFile(res, cfg.OutputToml)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func testSinglePerformance(ctx context.Context, queuedTestCases []testCaseName, allTestCases map[testCaseName]func(context.Context, *testPerformanceConfig) testResult, cfg testPerformanceConfig, resCh chan map[string][]testResult) {
+	defer close(resCh)
+	singleTestResCh := make(chan testResult)
+	allTestRes := []testResult{}
+	// run all performance tests for a performance client, pushing each completed test to the channel until all are complete or timeout occurs
+	go testPerformance(ctx, queuedTestCases, allTestCases, cfg, singleTestResCh)
+
+	testCounter := 0
+	finished := false
+	for !finished {
+		var testName string
+		select {
+		case <-ctx.Done():
+			testName = queuedTestCases[testCounter].name
+			allTestRes = append(allTestRes, testResult{Name: testName, Verdict: testVerdictFail, Error: errTimeoutInterrupted})
+			finished = true
+		case result, ok := <-singleTestResCh:
+			if !ok {
+				finished = true
+				break
+			}
+			testName = queuedTestCases[testCounter].name
+			testCounter++
+			result.Name = testName
+			allTestRes = append(allTestRes, result)
+		}
+	}
+
+	resCh <- map[string][]testResult{"local": allTestRes}
+}
+
+func testPerformance(ctx context.Context, queuedTests []testCaseName, allTests map[testCaseName]func(context.Context, *testPerformanceConfig) testResult, cfg testPerformanceConfig, ch chan testResult) {
+	defer close(ch)
+	for _, t := range queuedTests {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			ch <- allTests[t](ctx, &cfg)
+		}
+	}
+}
+
+func performanceDiskWriteTest(ctx context.Context, conf *testPerformanceConfig) testResult {
+	testRes := testResult{Name: "DiskWrite"}
+
+	log.Info(ctx, "Testing disk write...",
+		z.Any("file size MB", conf.DiskWriteMB),
+		z.Any("loops", diskWriteLoops))
+
+	if conf.DiskWriteMB <= 3072 {
+		log.Warn(ctx, "File size used for tests of 3072MB or lower impacts the measured performance", nil)
+	}
+
+	var stat unix.Statfs_t
+	wd, _ := os.Getwd()
+	err := unix.Statfs(wd, &stat)
+	if err != nil {
+		return failedTestResult(testRes, err)
+	}
+	// Available blocks * size per block = available space in bytes
+	availableMB := int(stat.Bavail * uint64(stat.Bsize) / 1024 / 1024)
+	actualDiskWriteMB := conf.DiskWriteMB
+
+	for availableMB < actualDiskWriteMB {
+		log.Warn(ctx, fmt.Sprintf("Insufficient available disk space of %vMB, reducing the test size to %vMB. Note that this might result in slower write speed", availableMB, actualDiskWriteMB), nil)
+		actualDiskWriteMB /= actualDiskWriteMB
+		if actualDiskWriteMB == 0 {
+			return failedTestResult(testRes, errors.New("insufficient available disk space", z.Str("available_space", strconv.Itoa(availableMB)+"MB")))
+		}
+	}
+
+	var diskWriteTotal float64
+	for range diskWriteLoops {
```

Comment:
I've set this loop in order to get a bit more normalised data from the test and not to test concurrency. If I know my disk write speed is 1500MB/s and this test gives me 300MB/s result it would be misleading?

Reply:
No, you will sum up each thread performance and still get a value close to what's advertised by manufacturer.
Misleading would be if we pretend our disk can do 1500MB/s in one thread, but in real scenario it will be 300, because all stack components run concurrently.

Reply:
Hmmm yap, that's a good point. One thing that bothers me though is that with disk writes below 4GB I observed decreased write speeds. For 512MB writes my write performance was 700-1000MB/s, while with 4GB it was 1450-1550MB/s. Above 4GB it didn't make difference in speed.